lib/utils/spawn.js

```diff
@@ -0,0 +1,306 @@
+/*!
+ * spawn.js - promise and generator control flow for bcoin
+ * Originally based on yoursnetwork's "asink" module.
+ * Copyright (c) 2014-2016, Christopher Jeffrey (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+var utils = require('./utils');
+
+/**
+ * Execute an instantiated generator.
+ * @param {Generator} gen
+ * @returns {Promise}
+ */
+
+function exec(gen) {
+  return new Promise(function(resolve, reject) {
+    function step(value, rejection) {
+      var next;
+
+      try {
+        if (rejection)
+          next = gen.throw(value);
+        else
+          next = gen.next(value);
+      } catch (e) {
+        reject(e);
+        return;
+      }
+
+      if (next.done) {
+        resolve(next.value);
+        return;
+      }
+
+      if (!isPromise(next.value)) {
+        step(next.value, false);
+        return;
+      }
+
+      next.value.then(succeed, fail);
```

Comment:
`succeed` just passes the resolved value back into `step` which returns it to the generator (i.e. the next `yield`).